M2/Macaulay2/m2/Core.m2

```diff
@@ -116,6 +116,21 @@ needs = filename -> if not filesLoaded#?filename then load filename else (
      (filepath, filetime) := filesLoaded#filename;
      if filetime < fileTime filepath then load filepath)
 
+-----------------------------------------------------------------------------
+-- Setup persistent history
+-----------------------------------------------------------------------------
+historyFilename = "~/.Macaulay2-history"
+
+if not gotarg "--no-readline" then (
+    addStartFunction(() -> readHistory historyFilename);
+    -- TODO: find a better alternative to addEndFunction, because
+    -- exiting with Ctrl+D duplicates the last line of history file,
+    -- but if we use lineNumber-1, then exit and restart miss the first
+    -- entry from the current session. Moreover, maybe we shouldn't save
+    -- exit and restart in the history anyway.
```

Comment:
For consistency with Emacs, I think it would make sense to save `exit` and `restart`.

Reply:
Sure, that's fine, but I'd like to figure out a way to detect when we exit by Ctrl-D. Python, Julia, etc. handle this fine (and save "exit"), so I'm sure there's a way. Maybe `lineNumber` is not the right counter to use, or maybe it should be incremented by `exit` and `restart` and `end` before they evaluate?